streaming/cleanup.go

```diff
@@ -0,0 +1,315 @@
+			} else {
+				cleanupPubsubResources(ctx, pubsubResources, project)
+			}
+		}
+	}
+	if jobCleanupOptions.Monitoring {
+		//fetch monitoring resources
+		shardMonitoringResourcesList, err := FetchResources(ctx, migrationJobId, constants.MONITORING_RESOURCE, dataShardIds, project, instance)
+		if err != nil {
+			logger.Log.Debug(fmt.Sprintf("Unable to fetch shard monitoring resources for jobId: %s: %v\n", migrationJobId, err))
+		}
+		jobMonitoringResourcesList, err := FetchResources(ctx, migrationJobId, constants.AGG_MONITORING_RESOURCE, dataShardIds, project, instance)
+		if err != nil {
+			logger.Log.Debug(fmt.Sprintf("Unable to fetch aggregate monitoring resources for jobId: %s: %v\n", migrationJobId, err))
+		}
+		monitoringResourcesList := append(shardMonitoringResourcesList, jobMonitoringResourcesList...)
+		//cleanup
+		for _, resources := range monitoringResourcesList {
+			var monitoringResources internal.MonitoringResources
+			err := json.Unmarshal([]byte(resources.ResourceMetadata), &monitoringResources)
+			if err != nil {
+				logger.Log.Debug("Unable to read monitoring metadata for deletion\n")
+			} else {
+				cleanupMonitoringDashboard(ctx, monitoringResources, project)
+			}
+		}
+	}
+}
+
+func FetchResources(ctx context.Context, migrationJobId string, resourceType string, dataShardIds []string, project string, instance string) ([]JobResources, error) {
+	dbURI := fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, constants.METADATA_DB)
+	client, err := utils.GetClient(ctx, dbURI)
+	if err != nil {
+		err = fmt.Errorf("can't create client for db %s: %v", dbURI, err)
+		return nil, err
+	}
+	defer client.Close()
+	txn := client.ReadOnlyTransaction()
+	defer txn.Close()
+
+	//fetch all resources
+	var resourceQuery spanner.Statement
+	if dataShardIds != nil {
+		//query the provided data shards only
+		resourceQuery = spanner.Statement{
+			SQL: `SELECT 
+					ResourceId,
+					JobId,
+					DataShardId,
+					ExternalResourceId,
+					ResourceName,
+					ResourceType,
+					TO_JSON_STRING(ResourceMetadata) AS ResourceMetadata
+				FROM JobResources 
+				WHERE JobId = @migrationJobId and ResourceType = @resourceType and DataShardId IN UNNEST (@dataShardIds)`,
+			Params: map[string]interface{}{
+				"migrationJobId": migrationJobId,
+				"resourceType":   resourceType,
+				"dataShardIds":   dataShardIds,
+			},
+		}
+	} else {
+		//query all data shards
+		resourceQuery = spanner.Statement{
+			SQL: `SELECT 
+					ResourceId,
+					JobId,
+					DataShardId,
+					ExternalResourceId,
+					ResourceName,
+					ResourceType,
+					TO_JSON_STRING(ResourceMetadata) AS ResourceMetadata
+				FROM JobResources 
+				WHERE JobId = @migrationJobId and ResourceType = @resourceType`,
+			Params: map[string]interface{}{
+				"migrationJobId": migrationJobId,
+				"resourceType":   resourceType,
+			},
+		}
+	}
+	iter := txn.Query(ctx, resourceQuery)
+	jobResourcesList := []JobResources{}
+	for {
```

Comment:
A more "Go" way of doing this would be 

> for row,e := txn.Query(..) {
> ....
> }

or

>for iter.Next() {
>}

Reply:
Interesting, I have just used the existing reference in the code here -

https://github.com/GoogleCloudPlatform/spanner-migration-tool/blob/2fa927e9dd835261eb2899bb24ec13bd18467bf1/webv2/session/remote_session_store.go#L60